### ISSUE TITLE
Admin error due to ModelLogEntry after site setting model has been removed

### ISSUE THREAD

**Issue by th3hamm0r** [wagtail/wagtail#8243]
### Issue Summary

I've removed a custom site settings model (actually I've reverted the migration), which has been edited before, so the `ModelLogEntry` model contained an `wagtail.edit`-action entry.
Now the site history page in the CMS (/admin/reports/site-history/) has crashed with the following error:

    Traceback (most recent call last):
      File "/***/lib/python3.10/site-packages/asgiref/sync.py", line 451, in thread_handler
        raise exc_info[1]
      File "/***/lib/python3.10/site-packages/django/core/handlers/exception.py", line 38, in inner
        response = await get_response(request)
      File "/***/lib/python3.10/site-packages/django/core/handlers/base.py", line 233, in _get_response_async
        response = await wrapped_callback(request, *callback_args, **callback_kwargs)
      File "/***/lib/python3.10/site-packages/asgiref/sync.py", line 414, in __call__
        ret = await asyncio.wait_for(future, timeout=None)
      File "/usr/lib/python3.10/asyncio/tasks.py", line 408, in wait_for
        return await fut
      File "/***/lib/python3.10/site-packages/asgiref/current_thread_executor.py", line 22, in run
        result = self.fn(*self.args, **self.kwargs)
      File "/***/lib/python3.10/site-packages/asgiref/sync.py", line 455, in thread_handler
        return func(*args, **kwargs)
      File "/***/lib/python3.10/site-packages/sentry_sdk/integrations/django/views.py", line 67, in sentry_wrapped_callback
        return callback(request, *args, **kwargs)
      File "/***/decorators.py", line 14, in wrapper
        return view(request, *args, **kwargs)
      File "/***/decorators.py", line 23, in wrapper
        return view(request, *args, **kwargs)
      File "/***/urls.py", line 66, in decorated_view
        return view_func(request, *args, **kwargs)
      File "/***/lib/python3.10/site-packages/django/views/decorators/cache.py", line 44, in _wrapped_view_func
        response = view_func(request, *args, **kwargs)
      File "/***/lib/python3.10/site-packages/wagtail/admin/urls/__init__.py", line 125, in wrapper
        return view_func(request, *args, **kwargs)
      File "/***/lib/python3.10/site-packages/wagtail/admin/auth.py", line 174, in decorated_view
        response = view_func(request, *args, **kwargs)
      File "/***/lib/python3.10/site-packages/django/views/generic/base.py", line 70, in view
        return self.dispatch(request, *args, **kwargs)
      File "/***/lib/python3.10/site-packages/django/views/generic/base.py", line 98, in dispatch
        return handler(request, *args, **kwargs)
      File "/***/lib/python3.10/site-packages/wagtail/admin/views/reports/base.py", line 44, in get
        context['object_list'] = self.decorate_paginated_queryset(context['object_list'])
      File "/***/lib/python3.10/site-packages/wagtail/admin/views/reports/audit_logging.py", line 175, in decorate_paginated_queryset
        for log_entry, instance in log_entries:
      File "/***/lib/python3.10/site-packages/wagtail/core/models/audit_log.py", line 66, in with_instances
        model_instances = model.objects.in_bulk(object_ids)
    AttributeError: 'NoneType' object has no attribute 'objects'

The `django_content_type` table still contained the removed site settings model, so I thought I can fix this issue by removing it, so I used django's `./manage.py remove_stale_contenttypes` to remove the content type.

Now, the audit logging report view crashes again with the following error:

    Traceback (most recent call last):
      File "/***/lib/python3.10/site-packages/django/contrib/contenttypes/models.py", line 110, in get_for_id
        ct = self._cache[self.db][id]
    KeyError: None

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/***/lib/python3.10/site-packages/asgiref/sync.py", line 451, in thread_handler
        raise exc_info[1]
      File "/***/lib/python3.10/site-packages/django/core/handlers/exception.py", line 38, in inner
        response = await get_response(request)
      File "/***/lib/python3.10/site-packages/django/core/handlers/base.py", line 233, in _get_response_async
        response = await wrapped_callback(request, *callback_args, **callback_kwargs)
      File "/***/lib/python3.10/site-packages/asgiref/sync.py", line 414, in __call__
        ret = await asyncio.wait_for(future, timeout=None)
      File "/usr/lib/python3.10/asyncio/tasks.py", line 408, in wait_for
        return await fut
      File "/***/lib/python3.10/site-packages/asgiref/current_thread_executor.py", line 22, in run
        result = self.fn(*self.args, **self.kwargs)
      File "/***/lib/python3.10/site-packages/asgiref/sync.py", line 455, in thread_handler
        return func(*args, **kwargs)
      File "/***/lib/python3.10/site-packages/sentry_sdk/integrations/django/views.py", line 67, in sentry_wrapped_callback
        return callback(request, *args, **kwargs)
      File "/***/decorators.py", line 14, in wrapper
        return view(request, *args, **kwargs)
      File "/***/decorators.py", line 23, in wrapper
        return view(request, *args, **kwargs)
      File "/***/urls.py", line 66, in decorated_view
        return view_func(request, *args, **kwargs)
      File "/***/lib/python3.10/site-packages/django/views/decorators/cache.py", line 44, in _wrapped_view_func
        response = view_func(request, *args, **kwargs)
      File "/***/lib/python3.10/site-packages/wagtail/admin/urls/__init__.py", line 125, in wrapper
        return view_func(request, *args, **kwargs)
      File "/***/lib/python3.10/site-packages/wagtail/admin/auth.py", line 174, in decorated_view
        response = view_func(request, *args, **kwargs)
      File "/***/lib/python3.10/site-packages/django/views/generic/base.py", line 70, in view
        return self.dispatch(request, *args, **kwargs)
      File "/***/lib/python3.10/site-packages/django/views/generic/base.py", line 98, in dispatch
        return handler(request, *args, **kwargs)
      File "/***/lib/python3.10/site-packages/wagtail/admin/views/reports/base.py", line 44, in get
        context['object_list'] = self.decorate_paginated_queryset(context['object_list'])
      File "/***/lib/python3.10/site-packages/wagtail/admin/views/reports/audit_logging.py", line 175, in decorate_paginated_queryset
        for log_entry, instance in log_entries:
      File "/***/lib/python3.10/site-packages/wagtail/core/models/audit_log.py", line 65, in with_instances
        model = ContentType.objects.get_for_id(content_type_id).model_class()
      File "/***/lib/python3.10/site-packages/django/contrib/contenttypes/models.py", line 114, in get_for_id
        ct = self.get(pk=id)
      File "/***/lib/python3.10/site-packages/django/db/models/manager.py", line 85, in manager_method
        return getattr(self.get_queryset(), name)(*args, **kwargs)
      File "/***/lib/python3.10/site-packages/django/db/models/query.py", line 435, in get
        raise self.model.DoesNotExist(
    django.contrib.contenttypes.models.ContentType.DoesNotExist: ContentType matching query does not exist.

The problem occurs, because wagtail's ModelLogEntry-table still contains the log entries for the removed model, only the "content_type"-column is set to `NULL`.

I think wagtail at least needs to handle the case where a content type gets removed, but its log entries remain in the db.
I don't know why the `BaseLogEntry` uses `ondelete=models.SET_NULL` and not `models.CASCADE` [here](https://github.com/wagtail/wagtail/blob/main/wagtail/models/audit_log.py#L144)?
Does it make sense to keep the log entries, when the content type gets removed?

The first error could be handled too, e.g. by silently ignoring ContentTypes that don't return actual model classes [here](https://github.com/wagtail/wagtail/blob/main/wagtail/models/audit_log.py#L67).

### Steps to Reproduce

1. Start a new project with `wagtail start myproject`
2. Create a custom settings model (https://docs.wagtail.org/en/stable/reference/contrib/settings.html)
3. Edit those settings in the CMS
4. Remove the model, eg. by down-migrating the migration created at step 2

### Technical details

* Python version: 3.10
* Django version: 3.2.12
* Wagtail version: 2.16.1

**Comment by th3hamm0r** [comment]
After stumbling over this issue again, I really would like to make a PR to fix this.

I think the `content_type` field of `BaseLogEntry` should not allow null values and should cascade deletes of the related `ContentType`. The current, nullable relation does not make any sense to me, because ContentTypes cannot be deleted through the admin, it is a developer's decision to remove a model (and its history), so the developer should also cleanup the stale ContentType with `django-admin remove_stale_contenttypes` ([docs](https://docs.djangoproject.com/en/stable/ref/django-admin/#remove-stale-contenttypes)), which would then delete all related `BaseLogEntry`-items.

It seems like this nullable relation is based on django's `LogEntry` model, which also has a nullable `content_type` (see https://github.com/django/django/blob/241fe59b74bb6031fa644f3ad55e6ad6a9187510/django/contrib/admin/models.py#L48-L65), but in contrast to wagtail a cascading user relationship. For me it makes more sense to keep those logs if the user gets deleted and only remove those logs if the model/ContentType gets removed. The first part is already implemented by wagtail, as it displays a "system"-placeholder for deleted users.

I would like to change
https://github.com/wagtail/wagtail/blob/820c27fca9abec82406dad0ca6cbbf66c7e9ed7d/wagtail/models/audit_log.py#L142-L149

to
```python
    content_type = models.ForeignKey(
        ContentType,
        models.CASCADE,
        verbose_name=_("content type"),
        related_name="+",
    )
```

The only challenge with this change is, that we probably have to delete all NULL-entries in the created django migration, because active installations of wagtail will probably already contain such NULL-entries in the db (if you don't use the history-page, you won't notice the above error). Also adding a note in the docs somewhere, that developers should cleanup the content types of removed models with `django-admin remove_stale_contenttypes` ([docs](https://docs.djangoproject.com/en/stable/ref/django-admin/#remove-stale-contenttypes)) would be good.

What do you think?

**Comment by AlessioNar** [comment]
I am also encountering this error, for the same exact reason. How did you patch it?

**Comment by th3hamm0r** [comment]
@AlessioNar sorry for the late reply. We haven't patched the code, we've just removed all the log entries with a NULL-content_type manually from the db.

**Comment by th3hamm0r** [comment]
@zerolab @gasman Sorry to bump you both, but I've found your initial commit for the model logging (https://github.com/wagtail/wagtail/commit/d7d63c3ab1445629cc6bc03f46f40f8781ec00f0#diff-fccf759288bb9309dd5cfbe05ae13e59d091e625fe53e351bb9d909bb739b0b2) and I haven't found a better way to ask/discuss my suggested PR below (is there a slack-channel which I should use for such questions?).

I would like to fix that issue with a PR in 2 steps:
1. Change the code so it can handle deleted model classes without crashing, so this should avoid the `NoneType`-`AttributeError`. With this, a developer can safely remove a model from python, e.g. when switching between branches with diverging migrations, still being able to view the audit logging without a crash.
2. Make the `ForeignKey` to `ContentType` non-NULL and cascading (as suggested in https://github.com/wagtail/wagtail/issues/8243#issuecomment-1134527938), so if a developer finally decides to remove stale content types with `manage.py remove_stale_contenttypes`, it won't leave (zombie) log entries in the db, which make no sense, I think.

**Comment by th3hamm0r** [comment]
Errors fixed in #10130
